allennlp/semparse/worlds/atis_world.py

```diff
@@ -31,74 +59,182 @@ def __init__(self,
                  utterances: List[str],
                  tokenizer: Tokenizer = None,
                  database_directory: str = None) -> None:
+
+        self.all_tables = ALL_TABLES
+        self.tables_with_strings = TABLES_WITH_STRINGS
         if database_directory:
-            self.sql_table_context = SqlTableContext(ALL_TABLES,
-                                                     TABLES_WITH_STRINGS,
-                                                     database_directory)
+            self.database_directory = database_directory
+            self.connection = sqlite3.connect(self.database_directory)
+            self.cursor = self.connection.cursor()
+
+        self.sql_table_context = SqlTableContext(ALL_TABLES,
+                                                 TABLES_WITH_STRINGS,
+                                                 database_directory,
+                                                 utterances=utterances) if database_directory else None
+
+        self.grammar_dictionary = deepcopy(self.sql_table_context.grammar_dictionary)
 
         self.utterances: List[str] = utterances
         self.tokenizer = tokenizer if tokenizer else WordTokenizer()
         self.tokenized_utterances = [self.tokenizer.tokenize(utterance) for utterance in self.utterances]
-        valid_actions, linking_scores = self.init_all_valid_actions()
-        self.valid_actions: Dict[str, List[str]] = valid_actions
+        self.linked_entities = self.get_linked_entities()
 
+        self.valid_actions: Dict[str, List[str]] = self.update_valid_actions()
         # This has shape (num_entities, num_utterance_tokens).
+        entities, linking_scores = self.flatten_entities()
         self.linking_scores: numpy.ndarray = linking_scores
+        self.entities: List[str] = entities
         self.grammar_str: str = self.get_grammar_str()
         self.grammar_with_context: Grammar = Grammar(self.grammar_str)
 
+        if database_directory:
+            self.connection.close()
+
+
     def get_valid_actions(self) -> Dict[str, List[str]]:
         return self.valid_actions
 
-    def init_all_valid_actions(self) -> Tuple[Dict[str, List[str]], numpy.ndarray]:
-        """
-        We initialize the valid actions with the global actions. We then iterate through the
-        utterances up to and including the current utterance and add the valid strings.
-        """
+    def add_to_number_linking_scores(self,
+                                     all_numbers: Set[str],
+                                     number_linking_scores: Dict[str, Tuple[str, str, List[int]]],
+                                     get_number_linking_dict: Callable[[str, List[Token]],
+                                                                       Dict[str, List[int]]],
+                                     current_tokenized_utterance: List[Token],
+                                     nonterminal: str) -> None:
+        number_linking_dict: Dict[str, List[int]] = {}
```

Comment:
I'm finding myself trying to figure out what this logic is doing by reading the code, and it's a bit hard.  A high-level comment or docstring explaining what this method is doing would be good.  And the same thing goes for most of the internal methods on this class.